### ISSUE TITLE
MHW losing keyboard at the start of Safi with latest version

### ISSUE THREAD

**Issue by GoLD-ReaVeR** [Frogging-Family/wine-tkg-git#65]
Hi, as title states. I'm willing it has to do with me switching focus a few times around during the intro.

**Comment by Tk-Glitch** [comment]
Yeah that one is an issue the game has had since its first successful run with Wine, but it's quite random. I have noticed that *sometimes* hitting alt can get it back, but not always.

**Comment by GoLD-ReaVeR** [comment]
This particular issue wasn't found in the previous version that I got (5.9.r22 and before). At least not by me. Pressing any key on the keyboard is completely pointless by the looks of it. I even tried minimizing the window which is prone to crashing the renderer.

**Comment by GoLD-ReaVeR** [comment]
Issue is still not fixed and happens for alatreon as well.

**Comment by GoLD-ReaVeR** [comment]
It pertains to any alt-tabbing, don't know the exact conditions for reproduction.

**Comment by GoLD-ReaVeR** [comment]
I wanted to try a new build just now but I'm getting failure to compile. Here's the log:
[prepare.log](https://github.com/Frogging-Family/wine-tkg-git/files/4963297/prepare.log)

This buy is preventing me from upgrading to any latest version. That means I'm still stuck at proton-tkg-git 5.9.r22.g7b78338b-1.

**Comment by Tk-Glitch** [comment]
The patching failure was fixed with https://github.com/Frogging-Family/wine-tkg-git/commit/8f97980c74895eb04f021b02b2a00c6dd9e4fbee

Someone reported a temporarily introduced alt-tabbing issue went away recently, so it might indeed be of interest to give it another try.

I wasn't able to reproduce your specific issue on my end in a consistent way.
The game has always been extremely playful with alt-tabbing on my end and not giving back inputs has always happened for me, rarely, on both linux and windows (and way before Iceborne). No change in behavior with Alatreon for me either. So there might be some DE/display mode greatness involved. I'm always playing in borderless myself for a start. Fullscreen mode used to crash the game randomly when switching focus. Not sure if it helps in any way :frog:

**Comment by GoLD-ReaVeR** [comment]
Also, did you manage to solo Alatreon? :D

**Comment by Tk-Glitch** [comment]
No yet :D But *I will succeed* - at some point -

**Comment by GoLD-ReaVeR** [comment]
I tried to build the latest version ~6 hours ago and the game didn't start. Found no indicators in the log, it just stops.

**Comment by GoLD-ReaVeR** [comment]
Here's the logs:
[steam-582010.log](https://github.com/Frogging-Family/wine-tkg-git/files/4969931/steam-582010.log)

**Comment by Tk-Glitch** [comment]
I have tested the game on a fresh build and faced no issue. It could have been bad timing with fixes not yet merged.. Or something breaking on the makepkg side. I'll try a makepkg build for a check.

**Comment by GoLD-ReaVeR** [comment]
Still not working for me. I tried both MHW and Guilty Gear Xrd Revelator and both crash before finishing any form of start up. The games don't even get to complain about anything, they just don't start.

**Comment by Tk-Glitch** [comment]
I can't reproduce the issue with a makepkg build either. Have you tried starting steam from term to see if anything pops up on game launch?

> I'm not getting a winecfg upgrade notification when switching to the new version.

That's expected unless you have added https://github.com/Frogging-Family/community-patches/blob/master/wine-tkg-git/unhide-prefix-update-window.mypatch to your community patches array. It's been set to hidden by default *kinda recently* to be in line with vanilla proton.

**Comment by GoLD-ReaVeR** [comment]
Traceback (most recent call last):
  File "/usr/share/steam/compatibilitytools.d/proton_tkg_makepkg/proton", line 875, in <module>
    g_session.init_session()
  File "/usr/share/steam/compatibilitytools.d/proton_tkg_makepkg/proton", line 673, in init_session
    g_compatdata.setup_prefix()
  File "/usr/share/steam/compatibilitytools.d/proton_tkg_makepkg/proton", line 485, in setup_prefix
    try_copy(g_proton.lib64_dir + "wine/vkd3d-proton/d3d12.dll",
  File "/usr/share/steam/compatibilitytools.d/proton_tkg_makepkg/proton", line 59, in try_copy
    shutil.copy(src, dst)
  File "/usr/lib/python3.8/shutil.py", line 415, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.8/shutil.py", line 261, in copyfile
    with open(src, 'rb') as fsrc, open(dst, 'wb') as fdst:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/share/steam/compatibilitytools.d/proton_tkg_makepkg/dist/lib64/wine/vkd3d-proton/d3d12.dll'
Game removed: AppID 582010 "", ProcID 831989 

Did something change in the config?

**Comment by GoLD-ReaVeR** [comment]
Ah, that rename in 021e1541a9099831ba5ce8fe843f6e6d470cdbea got undone with the merge.

**Comment by GoLD-ReaVeR** [comment]
Even when changing the variable in the config to _use_vkd3dlib="fork" the same error keeps happening.

**Comment by GoLD-ReaVeR** [comment]
I saw the updated and tried to recompile but I'm getting patch failures now.

**Comment by Tk-Glitch** [comment]
Staging was updated in the meantime :p
It hopefully should be all good now, even with `fork` or `mainline`. For that folder to be found, `false` was required, which requires mingw in return. So I can see how it was so problematic. For some reason I believed python's `try_copy` would fail gracefully in a non-blocking way. I was wrong.

> Also, I must point out that these comments above the settings are not helping either :P

True :sweat_smile: 
I'll have to take a few decisions regarding vkd3d implementation at some point so that'll get reviewed sooner than later.

**Comment by GoLD-ReaVeR** [comment]
Alright, that was the fastest crash to desktop in a while. I switched from gathering hub to seliana and moved focus to another window during the loadingscreen. I lost the ability to do anything and when I switched apps again the game crashed. (This is with dx12)

**Comment by GoLD-ReaVeR** [comment]
127299 538468.552:00b8:00cc:err:ntdll:RtlpWaitForCriticalSection section 0x7fc431041e00 "../../../wine-mirror-git/dlls/dinput8/../dinput/dinput_main.c: dinput_hook_crit" wait timed out in thread 00cc, blocked by 0198, retrying (60        sec)
127300 538533.553:00b8:00cc:err:ntdll:RtlpWaitForCriticalSection section 0x7fc431041e00 "../../../wine-mirror-git/dlls/dinput8/../dinput/dinput_main.c: dinput_hook_crit" wait timed out in thread 00cc, blocked by 0198, retrying (60        sec)

Well hello!

The log is full with a shitton of d3d_command_list_QueryInterface warnings. But the critical section may be the issue here. The full log is 127301 lines. Want me to send it?

**Comment by Tk-Glitch** [comment]
Yes please. As well as your `last_build_config.log`

**Comment by GoLD-ReaVeR** [comment]
[MHWLog.zip](https://github.com/Frogging-Family/wine-tkg-git/files/4995230/MHWLog.zip)

**Comment by Tk-Glitch** [comment]
Thank you !
Can I also ask you which DE you're using?

**Comment by GoLD-ReaVeR** [comment]
DE?

**Comment by Tk-Glitch** [comment]
Desktop Environment

**Comment by GoLD-ReaVeR** [comment]
I switch out of the game by using windows key + left/right to another desktop with another application running such as discord or webbrowser. The game is in windowed fullscreen as regular fullscreen is just a mess these days.

**Comment by Tk-Glitch** [comment]
https://wiki.archlinux.org/index.php/Desktop_environment

**Comment by Tk-Glitch** [comment]
Oh I didn't see your edit

**Comment by Tk-Glitch** [comment]
> awesomewm
considering the endless amount of issues with wine + tilings wms it's not a good start :frog:

**Comment by GoLD-ReaVeR** [comment]
How can anyone live without a tiling WM?

I haven't added a floating rule for wine/proton, it typically seems to float automatically.

**Comment by Tk-Glitch** [comment]
Small chat earlier today with someone encountering a very similar issue with Overwatch on a now pretty old build:
![Screenshot_20200730_235225](https://user-images.githubusercontent.com/741977/88978436-01902880-d2c0-11ea-861d-73e0781889b8.png)

**Comment by GoLD-ReaVeR** [comment]
Any update?

It seems that you have to focus another window for this problem to trigger.

**Comment by Tk-Glitch** [comment]
I have a couple ideas but I'll need more time. I'm away from home for another week then I'll be able to properly get back to it.

**Comment by GoLD-ReaVeR** [comment]
One more update from me, this problem seems to stem from the 5.0 release. I have focusing issues with proton 5.0+ (including Valve version) preventing me from focusing other windows, even in other games such as Guilty Gear. So I think something changed between 4.* and 5* that received a few fixes recently and now screws up completely when focus changes.

**Comment by Tk-Glitch** [comment]
I remember there were a few issues reported regarding the FS hack and Proton 5.0+, which would also affect us as a result. You're probably onto something.

**Comment by GoLD-ReaVeR** [comment]
Any updates now? :P

**Comment by GoLD-ReaVeR** [comment]
Hmm, I installed latest vkd3d and proton (both tkg) and the renderer glitched and eventually froze. Should I make a new issue for this?

**Comment by GoLD-ReaVeR** [comment]
I'm going to confirm the keyboard issue as a generic 5.0 input issue. While devices may not necessarily be lost, the games have random delays when receiving inputs which is giving me trouble with Guilty Gear (1 frame links) and it also pertains to hunts like safi where the character does extra unrequested movement during judgement. It's like the input lock is shared with something that is directly affected by the game's performance or other external factors. The input system as such is not fit for purpose in proton 5.*.

**Comment by Tk-Glitch** [comment]
Still not sure about the exact breaking point, but after a few chats it's getting closer. Could you please confirm it also happens to you with FS hack disabled?
Regarding `I installed latest vkd3d and proton (both tkg) and the renderer glitched and eventually froze` there was apparently a regression in vkd3d-proton + nvidia that may already be fixed.

**Comment by GoLD-ReaVeR** [comment]
Sorry, FS Hack? 

Besides stracker's I don't use any special tools for either program I mentioned. I run the game on a floating window desktop and when I switch away from the game I do so by switching desktops.

**Comment by Tk-Glitch** [comment]
> Sorry, FS Hack?

Proton's fullscreen hack (enabled with `_proton_fs_hack="true"`, which is default). It seems to have regressed since Proton 5.0 and is currently the main suspect.

**Comment by GoLD-ReaVeR** [comment]
I saw some settings were moved around, trying to recompile and I hope I got it all right. dxvk on git, wine_dxgi disabled and vkd3d set to fork.

**Comment by GoLD-ReaVeR** [comment]
latest vkd3d apparently still locks up my graphics card. I'm again reverting that. Disabling the fs hack seems to resolve a slow of issues related to input. Delays are gone, mouse feels more responsive. I also am noticing improved fps, though that may be related to vkd3d. I'll keep you posted.

**Comment by GoLD-ReaVeR** [comment]
Reverting vkd3d to a build I made in june seems to have fixed the crashing. Thus far no input issues, the game works better than ever before. To be clear, I disabled both fs hack and compositor bypass.

**Comment by Tk-Glitch** [comment]
Great news! The compositor bypass is not needed if you either don't use a compositor or you have a way to disable it for fullscreen windows. The spinning camera used to be an issue with some gamepads, so if you have one plugged in, it might be worth attempting to launch the game with it unplugged.

> dxvk on git, dxvk_dxgi disabled and vkd3d set to fork

It's perfect except for vkd3d. You'll want "false" these days for the best vkd3d support (it skips using the native library and builds a windows dll instead, offering a few more features than the native lib). That being said, I don't think MHW needs it. Feel free to experiment nonetheless :)

**Comment by GoLD-ReaVeR** [comment]
Hmmm, having fs_hack+compositor bypass disabled seems to cause Guilty Gear to turn off my main monitor. I'm guessing it's requesting a resolution that doesn't exist. It also has graphical glitches that have been in there since 5.0 (Jack-O's command throw causing a black flash after it slams the opponent into the ground), but that's of secondary importance. Which of the two settings is responsible for the resolution change being prevented? And is there maybe a way around this?

**Comment by GoLD-ReaVeR** [comment]
Joystick not working on latest build. Doesn't seem to respond to axes.

**Comment by Tk-Glitch** [comment]
This was reported in #134 and is a tracked upstream regression. I have since reverted the offending patchset in the hotfixer so it should be good on current master.

**Comment by GoLD-ReaVeR** [comment]
[324498.345521] NVRM: GPU at PCI:0000:09:00: GPU-21589442-001b-4b23-9b0e-073213285a8d
[324498.345524] NVRM: GPU Board Serial Number: 
[324498.345527] NVRM: Xid (PCI:0000:09:00): 31, pid=632327, Ch 0000003e, intr 10000000. MMU Fault: ENGINE GRAPHICS GPCCLIENT_T1_9 faulted @ 0xff_41e24000. Fault is of type FAULT_PTE ACCESS_TYPE_READ

vkd3d was set to false rather than fork and it crashed the moment I entered Seliana. It's starting to show a similar bug to dxvk now.

**Comment by Tk-Glitch** [comment]
Looks like either a nvidia or a vkd3d regression. People are reverting (4) commits to play horizon zero dawn lately, it might be related:
https://github.com/ValveSoftware/Proton/issues/4125#issuecomment-691812754

**Comment by Tk-Glitch** [comment]
Sorry for the delay, there was some pending refactoring making an answer I'd have given at the time not very useful. Building vkd3d-proton (standalone) is now possible with https://github.com/Frogging-Family/dxvk-tools , which supports userpatches, making reverts easier. I'll make one for you to revert the 4 commits listed in the proton issue.

**Comment by Tk-Glitch** [comment]
[allocator_reverts.vkd3drevert.txt](https://github.com/Frogging-Family/wine-tkg-git/files/5366342/allocator_reverts.vkd3drevert.txt)
Here you go.

Remove the trailing `.txt` and drop it inside the `dxvk-tools/VKD3D_PROTONBUILD/patches/` dir. You'll want to use an older commit for the reverts to apply cleanly. I have tested with `7238802806688ea4ec21f4ef0f0414e7c6c03991` myself.
To use it, edit the `upvkd3d-proton.cfg` in dxvk-tools root so that `CUSTOM_COMMIT=1` and `VKD3D_PROTON_COMMIT=7238802806688ea4ec21f4ef0f0414e7c6c03991`.
You can then run `./upvkd3d-proton build` to build your custom vkd3d-proton standalone dlls.

**Comment by GoLD-ReaVeR** [comment]
I seem to be slightly stupid. I built the dlls, put the 64 bit one in the MHW folder, set the dll load to native, builtin in winecfg and the logs keep telling me I'm loading builtin. Do I need to change the wine build?

**Comment by Tk-Glitch** [comment]
Most (if not all?) d3d12 games seem to use d3d12.dll from system32/sysWOW64 exclusively, so you'll want to paste your build in the`dist/lib64/wine/vkd3d-proton/` dir of your proton build (64-bit) and `dist/lib/wine/vkd3d-proton/` (32-bit), so it gets grabbed by the proton script.
Alternatively, you can do it with upvkd3d-proton:
```
# PROTON UPDATE WITH : ./upvkd3d-proton proton-dist - Requires _proton_dist_path to be set in upvkd3d-proton.cfg
```

**Comment by GoLD-ReaVeR** [comment]
I haven't put the dll in there at this time because the game hasn't crashed with the proton compile I made. Has there been any work on the fullscreen hack?

**Comment by Tk-Glitch** [comment]
Kinda, but it's a huge mess currently due to large changes upstream. GE and I are both waiting for the patchset to be rebased against 6.0 until doing further work on it.

I have played MHW for quite a few hours lately and faced no issue with inputs or anything really, so I'm closing this :frog:
